BaseLib/FileTools.cpp

```diff
@@ -37,6 +40,112 @@ bool IsFileExisting(const std::string &strFilename)
+                  pos_curly_brace_close - (pos_curly_brace_open + 1)),
+        pos_curly_brace_open, pos_curly_brace_close);
+}
+
+std::string containsKeyword(std::string const& str, std::string const& keyword)
+{
+    auto const position = str.find(keyword);
+    if (position != std::string::npos)
+    {
+        return str.substr(0, position);
+    }
+    return "";
+}
+
+template <typename T>
+bool substituteKeyword(std::string& result, std::string& parenthesized_string,
+                       std::string::size_type begin, std::string::size_type end,
+                       std::string const& keyword, T& data)
+{
+    std::string precision_specification =
+        containsKeyword(parenthesized_string, keyword);
+
+    if (precision_specification.empty())
+    {
+        return false;
+    }
+
+    std::unordered_map<std::type_index, char> type_specification;
+    type_specification[std::type_index(typeid(int))] = 'd';
+    type_specification[std::type_index(typeid(double))] = 'f'; // default
+    type_specification[std::type_index(typeid(std::string))] = 's';
+
+    auto const& b = precision_specification.back();
+    // see https://fmt.dev/latest/syntax.html#formatspec
+    if (b == 'e' || b == 'E' || b == 'f' || b == 'F' || b == 'g' || b == 'G')
+    {
+        type_specification[std::type_index(typeid(double))] =
+            precision_specification.back();
+        precision_specification = precision_specification.substr(
+            0, precision_specification.length() - 1);
+    }
+
+    std::string const generated_fmt_string =
+        "{" + precision_specification +
+        type_specification[std::type_index(typeid(data))] + "}";
+    result = result.substr(0, begin) + fmt::format(generated_fmt_string, data) +
+             result.substr(end + 1, result.length() - (end + 1));
+    return true;
+}
+
+std::string constructFormattedFileName(std::string const& format_specification,
+                                       std::string const& mesh_name,
+                                       int const process_id,
+                                       int const timestep,
+                                       double const t)
+{
+    char const open_char = '{';
+    char const close_char = '}';
+    std::string::size_type begin = 0;
+    std::string::size_type end = std::string::npos;
+    std::string result = format_specification;
+
+    while (begin != std::string::npos)
+    {
+        auto length_before_substitution = result.length();
+        // find next parenthesized string
+        std::string str = "";
+        std::tie(str, begin, end) =
+            getParenthesizedString(result, open_char, close_char, begin);
+        if (!substituteKeyword(result, str, begin, end, "timestep", timestep))
+        {
+            if (!substituteKeyword(result, str, begin, end, "time", t))
+            {
+                if (!substituteKeyword(result, str, begin, end, "process_id",
+                                       process_id))
```

Comment:
Can this be reorganized in more readable style?
I think these are equivalent:
```c++
if (!A) {
 if (!B) {
  if (!C) {
   do()
}}}
```
```c++
if (!A && !B && !C) { do(); }
```
```c++
if (A || B || C) { do(); }  // This is my favorite ;)
```
```c++
if (A) {
  do();
  continue;
}
if (B) {
   do();
   continue;
}
...
```

Reply:
Is your favorite equivalent to the previous versions? IMHO it have to be
```C++
if (!(A || B || C)) { do(); }
```
In this case I prefer your second version.

Reply:
Sure. Thanks for correction.